plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,207 @@
+   snapshot_requests _snapshot_requests;
+   snapshot_db_json _snapshot_db;
+   uint32_t _snapshot_id   = 0;
+   uint32_t _inflight_sid  = 0;
+   std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> _create_snapshot;
+
+   void x_serialize() {
+      auto& vec = _snapshot_requests.get<as_vector>();
+      std::vector<producer_plugin::snapshot_schedule_information> sr(vec.begin(), vec.end());
+      _snapshot_db << sr;   
+   }
+
+public:
+   snapshot_scheduler() = default;
+
+   // snapshot_scheduler_listener
+   void on_start_block(uint32_t height) {
+      bool serialize_needed  = false;
+      bool snapshot_executed = false; 
+
+      auto execute_snapshot_with_log = [this, &height, &snapshot_executed](const auto & req) {
+         // one snapshot per height
+         if (!snapshot_executed) {
+            dlog("snapshot scheduler creating a snapshot from the request [start_block_num:${start_block_num}, end_block_num=${end_block_num}, block_spacing=${block_spacing}], height=${height}",
+               ("start_block_num", req.start_block_num)
+               ("end_block_num",   req.end_block_num)
+               ("block_spacing",   req.block_spacing)
+               ("height",          height));
+               
+            execute_snapshot(req.snapshot_request_id);
+            snapshot_executed = true;
+         }
+      };    
+
+      for(const auto& req: _snapshot_requests.get<0>()) {
+         // -1 since its called from start block
+         bool recurring_snapshot =  req.block_spacing &&  (height >= req.start_block_num + 1) && (!((height - req.start_block_num - 1) % req.block_spacing));
+         bool onetime_snapshot   = (!req.block_spacing) && (height == req.start_block_num + 1);
+        
+         // assume "asap" for snapshot with missed/zero start, it can have spacing
+         if(!req.start_block_num) {
+            // update start_block_num with current height only if this is recurring
+            // if non recurring, will be executed and unscheduled
+            if (req.block_spacing && height) {
+               auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+               auto it = snapshot_by_id.find(req.snapshot_request_id);
+               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height - 1; });
+               serialize_needed = true;              
+            }
+            execute_snapshot_with_log(req);
+         }        
+         else if(recurring_snapshot || onetime_snapshot) {
+            execute_snapshot_with_log(req);
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num ||         
+            (!req.block_spacing && height >= (req.start_block_num + 1)) || 
+            (req.end_block_num > 0 && height >= (req.end_block_num + 1))) {
+            unschedule_snapshot(req.snapshot_request_id);
+            if (snapshot_executed) break;
```

Comment:
This `if` statement should be one line further (after the closing brace).

Reply:
This way it will break loop that may have another start_block_num's to correct... It needs to be completed fully and block for execution of snapshot is now in execute_snapshot_with_log lambda. What do you think?

Reply:
Yes, but we don't want to take two snapshots at the same block height.
In any case, if we already too a snapshot, it is OK to update the `start_block_num` the next time this function gets called (at height+1).

Reply:
Oh, yes, right, I missed that. So we probably can remove line 99, right?

Reply:
done! thanks again!